File: src/lib/xdo.c
#include "init.h"
#include <xdo.h>
#include <stdio.h>
#include <sys/types.h>
#include <signal.h>
#include <unistd.h>

Window win;

int get_active_window() {
  xdo_get_active_window(xdo, &win);
  return win;
}

void move_mouse(int x, int y, int screen) {
  xdo_move_mouse(xdo, x, y, screen);
}

void move_mouse_relative_to_window(Window window, int x, int y) {
  xdo_move_mouse_relative_to_window(xdo, window, x, y);
}

void move_mouse_relative(int x, int y) {
  xdo_move_mouse_relative(xdo, x, y);
}

void mouse_down(Window window, int button) {
  xdo_mouse_down(xdo, window, button);
}

void mouse_up(Window window, int button) {
  xdo_mouse_up(xdo, window, button);
}

void get_mouse_location(int* buffer) {
  xdo_get_mouse_location(xdo, &buffer[0], &buffer[1], &buffer[2]);
}

Window get_window_at_mouse() {
  int ret = xdo_get_window_at_mouse(xdo, &win);
  if (ret != 0) {
    return 0;
  }
  return win;
}

void wait_for_mouse_move_from(int x, int y) {
  xdo_wait_for_mouse_move_from(xdo, x, y);
}

void wait_for_mouse_move_to(int x, int y) {
  xdo_wait_for_mouse_move_to(xdo, x, y);
}

void click_window(Window window, int button) {
  xdo_click_window(xdo, window, button);
}

void click_window_multiple(Window window, int button, int repeat, unsigned int delay) {
  xdo_click_window_multiple(xdo, window, button, repeat, delay);
}

void enter_text_window(Window window, char* text, unsigned int delay) {
  xdo_enter_text_window(xdo, window, text, delay);
}

void send_keysequence_window(Window window, char* keysequence, unsigned int delay) {
  xdo_send_keysequence_window(xdo, window, keysequence, delay);
}

void send_keysequence_window_up(Window window, char* keysequence, unsigned int delay) {
  xdo_send_keysequence_window_up(xdo, window, keysequence, delay);
}

void send_keysequence_window_down(Window window, char* keysequence, unsigned int delay) {
  xdo_send_keysequence_window_down(xdo, window, keysequence, delay);
}

void move_window(Window window, int x, int y) {
  xdo_move_window(xdo, window, x, y);
}

void set_window_size(Window window, int w, int h, int flags) {
  xdo_set_window_size(xdo, window, w, h, flags);
}

void set_window_property(Window window, char* key, char* value) {
  xdo_set_window_property(xdo, window, key, value);
}

void set_window_class(Window window, char* name, char* class) {
  xdo_set_window_class(xdo, window, name, class);
}

void set_window_urgency(Window window, int urgency) {
  xdo_set_window_urgency(xdo, window, urgency);
}

void focus_window(Window window) {
  xdo_activate_window(xdo, window);
}

void raise_window(Window window) {
  xdo_raise_window(xdo, window);
}

void activate_window(Window window) {
  xdo_activate_window(xdo, window);
}

Window get_focused_window() {
  xdo_get_focused_window(xdo, &win);
  return win;
}

void wait_for_window_focus(Window window) {
  xdo_wait_for_window_focus(xdo, window, 0);
}

int get_window_pid(Window window) {
  return xdo_get_pid_window(xdo, window);
}

void get_window_location(Window window, int* buffer) {
  xdo_get_window_location(xdo, window, &buffer[0], &buffer[1], NULL);
}

void get_window_size(Window window, unsigned int* buffer) {
  xdo_get_window_size(xdo, window, &buffer[0], &buffer[1]);
}

Window wait_select_window_with_click() {
  xdo_select_window_with_click(xdo, &win);
  return win;
}

unsigned int get_input_state() {
  return xdo_get_input_state(xdo);
}

void close_window(Window window) {
  xdo_close_window(xdo, window);
}

void kill_window(Window window) {
  xdo_kill_window(xdo, window);
}

void quit_window(Window window) {
  xdo_quit_window(xdo, window);
}

unsigned char* get_window_name(Window window) {
  unsigned char* name_ret;
  int name_len_ret;
  int name_type;
  int ret = xdo_get_window_name(xdo, window, &name_ret, &name_len_ret, &name_type);
  return name_ret;
}

File: src/lib/xdo.zig
// This file just wraps libxdo for the most part
const std = @import("std");
const global = @import("./global.zig");
const Window = X.Window;
const X = global.X;
//

pub fn moveMouse(x: i32, y: i32, screen: i32) callconv(.C) void {
    _ = X.xdo_move_mouse(global.xdo, x, y, screen);
}

pub fn getActiveWindow() callconv(.C) Window {
    var window: Window = 0;
    _ = X.xdo_get_active_window(global.xdo, &window);
    return window;
}

pub fn moveMouseRelativeToWindow(w: Window, x: i32, y: i32) callconv(.C) void {
    _ = X.xdo_move_mouse_relative_to_window(global.xdo, w, x, y);
}

pub fn moveMouseRelative(x: i32, y: i32) callconv(.C) void {
    _ = X.xdo_move_mouse_relative(global.xdo, x, y);
}

pub fn mouseDown(w: Window, button: i32) callconv(.C) void {
    _ = X.xdo_mouse_down(global.xdo, w, button);
}

pub fn mouseUp(w: Window, button: i32) callconv(.C) void {
    _ = X.xdo_mouse_up(global.xdo, w, button);
}

pub fn getMouseLocation(buffer: [*]i32) callconv(.C) void {
    _ = X.xdo_get_mouse_location(global.xdo, &buffer[0], &buffer[1], &buffer[2]);
}

pub fn getWindowAtMouse() callconv(.C) Window {
    var window: Window = undefined;
    if (X.xdo_get_window_at_mouse(global.xdo, &window) != 0) {
        return window;
    }
    return 0;
}

pub fn getWindowName(w: Window) callconv(.C) [*]const u8 {
    var name: [*c]u8 = undefined;
    var name_len: i32 = undefined;
    var name_type: i32 = undefined;
    _ = X.xdo_get_window_name(global.xdo, w, &name, &name_len, &name_type);
    return name;
}

pub fn waitForMouseMoveFrom(x: i32, y: i32) callconv(.C) void {
    _ = X.xdo_wait_for_mouse_move_from(global.xdo, x, y);
}

pub fn waitForMouseMoveTo(x: i32, y: i32) callconv(.C) void {
    _ = X.xdo_wait_for_mouse_move_to(global.xdo, x, y);
}

pub fn clickWindow(w: Window, button: i32) callconv(.C) void {
    _ = X.xdo_click_window(global.xdo, w, button);
}

pub fn clickWindowMultiple(w: Window, button: i32, repeat: i32, delay: u32) callconv(.C) void {
    _ = X.xdo_click_window_multiple(global.xdo, w, button, repeat, delay);
}

pub fn enterTextWindow(w: Window, text: [*]const u8, delay: u32) callconv(.C) void {
    _ = X.xdo_enter_text_window(global.xdo, w, text, delay);
}

pub fn sendKeysequenceWindow(w: Window, keysequence: [*]const u8, delay: u32) callconv(.C) void {
    _ = X.xdo_send_keysequence_window(global.xdo, w, keysequence, delay);
}

pub fn sendKeysequenceWindowDown(w: Window, keysequence: [*]const u8, delay: u32) callconv(.C) void {
    _ = X.xdo_send_keysequence_window_down(global.xdo, w, keysequence, delay);
}

pub fn sendKeysequenceWindowUp(w: Window, keysequence: [*]const u8, delay: u32) callconv(.C) void {
    _ = X.xdo_send_keysequence_window_up(global.xdo, w, keysequence, delay);
}

pub fn moveWindow(w: Window, x: i32, y: i32) callconv(.C) void {
    _ = X.xdo_move_window(global.xdo, w, x, y);
}

pub fn setWindowSize(w: Window, width: i32, height: i32) callconv(.C) void {
    _ = X.xdo_set_window_size(global.xdo, w, width, height, 0);
}

pub fn setWindowProperty(w: Window, property: [*]const u8, value: [*]const u8) callconv(.C) void {
    _ = X.xdo_set_window_property(global.xdo, w, property, value);
}

pub fn setWindowClass(w: Window, name: [*]const u8, class: [*]const u8) callconv(.C) void {
    _ = X.xdo_set_window_class(global.xdo, w, name, class);
}

pub fn setWindowUrgency(w: Window, urgency: i32) callconv(.C) void {
    _ = X.xdo_set_window_urgency(global.xdo, w, urgency);
}

pub fn activateWindow(w: Window) callconv(.C) void {
    _ = X.xdo_activate_window(global.xdo, w);
}

pub fn focusWindow(w: Window) callconv(.C) void {
    _ = X.xdo_focus_window(global.xdo, w);
}

pub fn raiseWindow(w: Window) callconv(.C) void {
    _ = X.xdo_raise_window(global.xdo, w);
}

pub fn getFocusedWindow() callconv(.C) Window {
    var window: Window = undefined;
    if (X.xdo_get_focused_window(global.xdo, &window) != 0) {
        return window;
    }
    return 0;
}

pub fn getWindowPID(w: Window) callconv(.C) i32 {
    return X.xdo_get_pid_window(global.xdo, w);
}

pub fn getWindowLocation(w: Window, buffer: [*]i32) callconv(.C) void {
    _ = X.xdo_get_window_location(global.xdo, w, &buffer[0], &buffer[1], 0);
}

pub fn getWindowSize(w: Window, buffer: [*]u32) callconv(.C) void {
    _ = X.xdo_get_window_size(global.xdo, w, &buffer[0], &buffer[1]);
}

pub fn waitSelectWindowWithClick() callconv(.C) Window {
    var w: Window = undefined;
    _ = X.xdo_select_window_with_click(global.xdo, &w);
    return w;
}

pub fn getInputState() callconv(.C) u32 {
    return X.xdo_get_input_state(global.xdo);
}

// pub fn getSymbolMap() callconv(.C) [*][*]const u8 {
//     return X.xdo_get_symbol_map();
// }

pub fn killWindow(w: Window) callconv(.C) void {
    _ = X.xdo_kill_window(global.xdo, w);
}

pub fn closeWindow(w: Window) callconv(.C) void {
    _ = X.xdo_quit_window(global.xdo, w);
}

// Searching for one window is done without passing a buffer
pub fn searchWindowSingle(
    mask: u32,
    require: u32,
    class: [*c]const u8,
    classname: [*c]const u8,
    name: [*c]const u8,
    role: [*c]const u8,
    max_depth: i32,
    pid: i32,
    only_visible: bool,
) callconv(.C) Window {
    _ = only_visible;
    var windowList: [*c]Window = undefined;
    var n: u32 = undefined;
    var search = X.xdo_search{
        .title = null,
        .winclass = class,
        .winclassname = classname,
        .winname = name,
        .winrole = role,
        .max_depth = max_depth,
        .pid = pid,
        .only_visible = 1,
        .require = require,
        .searchmask = mask,
        .screen = 0,
        .desktop = 0,
        .limit = 1,
    };
    _ = X.xdo_search_windows(
        global.xdo,
        &search,
        &windowList,
        &n,
    );
    defer std.c.free(windowList);

    if (n > 0) {
        return windowList[0];
    } else {
        return 0;
    }
}

// When multiple windows are requests, we pass a buffer. Unfortunatly, we can't just have XDO
// write to that buffer, as they allocate their own. So this just simplifies it to copy the memory.
pub fn searchWindowMultiple(
    mask: u32,
    require: u32,
    class: [*]const u8,
    classname: [*]const u8,
    name: [*]const u8,
    role: [*]const u8,
    max_depth: i32,
    pid: i32,
    only_visible: bool,
    buffer: [*]Window,
    limit: u32,
) callconv(.C) u32 {
    _ = only_visible;
    var windowList: [*c]Window = undefined;
    var n: u32 = undefined;
    _ = X.xdo_search_windows(
        global.xdo,
        &X.xdo_search{
            .title = null,
            .winclass = class,
            .winclassname = classname,
            .winname = name,
            .winrole = role,
            .max_depth = max_depth,
            .pid = pid,
            .only_visible = 1,
            .require = require,
            .searchmask = mask,
            .screen = 0,
            .desktop = 0,
            .limit = limit,
        },
        &windowList,
        &n,
    );
    defer std.c.free(windowList);
    if (n > 0) {
        std.mem.copy(Window, buffer[0..n], windowList[0..n]);
        return n;
    } else {
        return 0;
    }
}
